api/institutions/authentication.py

```diff
@@ -64,35 +72,75 @@ def authenticate(self, request):
+        # Use given name and family name to build full name if it is not provided
         if given_name and family_name and not fullname:
             fullname = given_name + ' ' + family_name
 
-        # institution must provide `fullname`, otherwise we fail the authentication and inform sentry
+        # Non-empty full name is required. Fail the auth and inform sentry if not provided.
         if not fullname:
-            message = 'Institution login failed: fullname required' \
-                      ' for user {} from institution {}'.format(username, provider['id'])
+            message = 'Institution login failed: fullname required for ' \
+                      'user {} from institution {}'.format(username, provider['id'])
             sentry.log_message(message)
             raise AuthenticationFailed(message)
 
-        # `get_or_create_user()` guesses names from fullname
-        # replace the guessed ones if the names are provided from the authentication
+        # Get an existing user or create a new one. If a new user is created, the user object is
+        # confirmed but not registered,which is temporarily of an inactive status. If an existing
+        # user is found, it is also possible that the user is inactive (e.g. unclaimed, disabled,
+        # unconfirmed, etc.).
         user, created = get_or_create_user(fullname, username, reset_password=False)
-        if created:
+
+        # Existing but inactive users need to be either "activated" or failed the auth
+        activation_required = False
+        new_password_required = False
+        if not created:
+            try:
+                drf.check_user(user)
+            except exceptions.UnclaimedAccountError:
+                # Unclaimed user (i.e. a user that has been added as an unregistered contributor)
+                user.unclaimed_records = {}
+                activation_required = True
+                # Unclaimed users have an unusable password when being added as an unregistered
+                # contributor. Thus a random usable password must be assigned during activation.
+                new_password_required = True
+            except exceptions.UnconfirmedAccountError:
+                # Unconfirmed user (i.e. a user that has been created during sign-up)
+                user.email_verifications = {}
+                activation_required = True
+                # Unconfirmed users already have a usable password set by the creator during
+                # sign-up. However, it must be overwritten by a new random one so the creator
+                # (if he is not the real person) can not access the account after activation.
+                new_password_required = True
+            except exceptions.DeactivatedAccountError:
+                # Deactivated user: login is not allowed for deactivated users
+                logger.error('Can not log into a deactivated account via institution SSO')
```

Comment:
* Error messages are neither  displayed to the user nor be sent back to CAS since current CAS only cares for the status code. I agree that it would be better if CAS could understand the error messages and thus provide more relevant information back to the user about the failure. However, this requires CAS change.
* In addition, I will add sentry logs here so we can be notified to track the issue.